msbuild/Xamarin.MacDev.Tasks/Tasks/AnalyzeFileChanges.cs

```diff
@@ -0,0 +1,127 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Xamarin.Localization.MSBuild;
+using Xamarin.Messaging.Build.Client;
+
+namespace Xamarin.MacDev.Tasks {
+	public class AnalyzeFileChanges : XamarinTask, ITaskCallback {
+		[Required]
+		public string WorkingDirectory { get; set; } = string.Empty;
+
+		[Required]
+		public ITaskItem? ReportFile { get; set; }
+
+		[Output]
+		public ITaskItem [] ChangedFiles { get; set; } = [];
+
+		public IEnumerable<ITaskItem> GetAdditionalItemsToBeCopied () => [];
+
+		//We need the ReportFile input to be copied to the Mac. Since it's the only ITaskItem input, it's safe to return true
+		public bool ShouldCopyToBuildServer (ITaskItem item) => true;
+
+		//In case it's a remote execution, we don't want empty output files to be copied since we need the real files to be copied
+		public bool ShouldCreateOutputFile (ITaskItem item) => false;
+
+		public override bool Execute ()
+		{
+			if (ShouldExecuteRemotely ()) {
+				return new TaskRunner (SessionId, BuildEngine4).RunAsync (this).Result;
+			}
+
+			if (!Directory.Exists (WorkingDirectory)) {
+				Log.LogError (MSBStrings.E7146 /* Unable to analyze file changes in '{0}'. Directory not found. */, WorkingDirectory);
+
+				return false;
+			}
+
+			if (!File.Exists (ReportFile!.ItemSpec)) {
+				Log.LogError (MSBStrings.E7147 /* Unable to analyze file changes in '{0}'. The report file '{1}' does not exist. */, WorkingDirectory, ReportFile.ItemSpec);
+
+				return false;
+			}
+
+			try {
+				var changedFiles = new List<ITaskItem> ();
+				//Gets a dictionary of file names, lengths and MVIDs from the ReportFile
+				IDictionary<string, (long length, Guid mvid)> reportFileList = GetReportFileList ();
+				IEnumerable<string> files = Directory.GetFiles (WorkingDirectory, "*.dll", SearchOption.TopDirectoryOnly);
+
+				foreach (string file in files) {
+					//We only want the files that are present in the ReportFile
+					if (!reportFileList.TryGetValue (Path.GetFileName (file), out (long length, Guid mvid) localInfo)) {
+						changedFiles.Add (new TaskItem (file));
+						TryAddPdbFile (file, changedFiles);
```

Comment:
I'm not sure I'm reading this right, but the comment doesn't match the code: aren't you adding files that are *not* present in the report file into the list of changed files here?

Reply:
Outdated comment, I'll remove it

Reply:
Done